examples/step-66/step-66.cc

```diff
@@ -0,0 +1,1156 @@
+
+    this->set_constrained_entries_to_one(inverse_diagonal);
+
+    for (unsigned int i = 0; i < inverse_diagonal.local_size(); ++i)
+      {
+        Assert(
+          inverse_diagonal.local_element(i) > 0.,
+          ExcMessage(
+            "No diagonal entry in a positive definite operator should be zero"));
+        inverse_diagonal.local_element(i) =
+          1. / inverse_diagonal.local_element(i);
+      }
+  }
+
+
+
+  template <int dim, int fe_degree, typename number>
+  void JacobianOperator<dim, fe_degree, number>::local_compute_diagonal(
+    const MatrixFree<dim, number> &             data,
+    LinearAlgebra::distributed::Vector<number> &dst,
+    const unsigned int &,
+    const std::pair<unsigned int, unsigned int> &cell_range) const
+  {
+    FEEvaluation<dim, fe_degree, fe_degree + 1, 1, number> phi(data);
+
+    AlignedVector<VectorizedArray<number>> diagonal(phi.dofs_per_cell);
+
+    for (unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
+      {
+        AssertDimension(nonlinear_values.size(0), data.n_macro_cells());
+        AssertDimension(nonlinear_values.size(1), phi.n_q_points);
+
+        phi.reinit(cell);
+        for (unsigned int i = 0; i < phi.dofs_per_cell; ++i)
+          {
+            for (unsigned int j = 0; j < phi.dofs_per_cell; ++j)
+              phi.submit_dof_value(VectorizedArray<number>(), j);
+            phi.submit_dof_value(make_vectorized_array<number>(1.), i);
+
+            phi.evaluate(true, true);
+            for (unsigned int q = 0; q < phi.n_q_points; ++q)
+              {
+                phi.submit_value(-nonlinear_values(cell, q) * phi.get_value(q),
+                                 q);
+                phi.submit_gradient(phi.get_gradient(q), q);
+              }
+            phi.integrate(true, true);
+            diagonal[i] = phi.get_dof_value(i);
+          }
+        for (unsigned int i = 0; i < phi.dofs_per_cell; ++i)
+          phi.submit_dof_value(diagonal[i], i);
+        phi.distribute_local_to_global(dst);
+      }
+  }
+
+
+
+  // @sect4{ResidualOperator}
+
+  // The next class implements the ResidualOperator, which we use to evaluate
+  // the residual and to assemble the right hand side. Since we will not pass
+  // objects of this class to any other objects for example linear solvers, etc.
+  // we have not need to fulfill any interface requirements. The operator's main
+  // functionality is implemented in the apply function. We can actually think
+  // of ResidualOperator.apply(dst, src) is mathematically expressed as
+  // F(src) = dst, where $F\colon\mathbb{R}^N\to\mathbb{R}^N$ is the discrete
+  // residual introduced in the introduction. Further the ResidualOperator has
+  // an initialization function, which stores a shared pointer to the MatrixFree
+  // object handling the loop over all cells, and a local_apply function
+  // implementing the calculation of the cell contribution.
+
+  // TODO: As functor with operator() function?
```

Comment:
I agree on that - I would remove the boilerplate of having a class and simply write the code inline where you use the residual with a functor - or add a function `residual` to the `JacobianOperator` class (well, then the name is not so clear).